src/nrniv/neuronapi.cpp

```diff
@@ -0,0 +1,524 @@
+}
+
+void nrn_section_pop(void) {
+    nrn_sec_pop();
+}
+
+void nrn_mechanism_insert(Section* sec, Symbol* mechanism) {
+    // TODO: throw exception if mechanism is not an insertable mechanism?
+    mech_insert1(sec, mechanism->subtype);
+}
+
+/****************************************
+ * Segments
+ ****************************************/
+
+int nrn_nseg_get(Section const* const sec) {
+    // always one more node than nseg
+    return sec->nnode - 1;
+}
+
+void nrn_nseg_set(Section* const sec, const int nseg) {
+    nrn_change_nseg(sec, nseg);
+}
+
+void nrn_segment_diam_set(Section* const sec, const double x, const double diam) {
+    Node* const node = node_exact(sec, x);
+    // TODO: this is fine if no 3D points; does it work if there are 3D points?
+    for (auto prop = node->prop; prop; prop = prop->next) {
+        if (prop->_type == MORPHOLOGY) {
+            prop->param(0) = diam;
+            diam_changed = 1;
+            node->sec->recalc_area_ = 1;
+            break;
+        }
+    }
+}
+
+double nrn_rangevar_get(const Symbol* sym, const Section* const sec, double x) {
+    return *nrn_rangepointer(const_cast<Section*>(sec), const_cast<Symbol*>(sym), x);
+}
+
+void nrn_rangevar_set(Symbol* sym, Section* const sec, double x, double value) {
+    *nrn_rangepointer(sec, sym, x) = value;
+}
+
+void nrn_rangevar_push(Symbol* sym, Section* const sec, double x) {
+    hoc_push(nrn_rangepointer(sec, sym, x));
+}
+
+nrn_Item* nrn_allsec(void) {
+    return static_cast<nrn_Item*>(section_list);
+}
+
+nrn_Item* nrn_sectionlist_data(Object* obj) {
+    // TODO: verify the obj is in fact a SectionList
+    return (nrn_Item*) obj->u.this_pointer;
+}
+
+/****************************************
+ * Functions, objects, and the stack
+ ****************************************/
+
+Symbol* nrn_symbol(char const* const name) {
+    return hoc_lookup(name);
+}
+
+int nrn_symbol_type(Symbol const* sym) {
+    // TODO: these types are in parse.hpp and are not the same between versions,
+    // so we really should wrap
+    return sym->type;
+}
+
+void nrn_symbol_push(Symbol* sym) {
+    hoc_pushpx(sym->u.pval);
+}
+
+/*double* nrn_get_symbol_ptr(Symbol* sym) {
+    return sym->u.pval;
+}*/
+
+void nrn_double_push(double val) {
+    hoc_pushx(val);
+}
+
+double nrn_double_pop(void) {
+    return hoc_xpop();
+}
+
+void nrn_double_ptr_push(double* addr) {
+    hoc_pushpx(addr);
+}
+
+double* nrn_double_ptr_pop(void) {
+    return hoc_pxpop();
+}
+
+void nrn_str_push(char** str) {
+    hoc_pushstr(str);
+}
+
+char** nrn_pop_str(void) {
+    return hoc_strpop();
+}
+
+void nrn_int_push(int i) {
+    hoc_pushi(i);
+}
+
+int nrn_int_pop(void) {
+    return hoc_ipop();
+}
+
+void nrn_object_push(Object* obj) {
+    hoc_push_object(obj);
+}
+
+Object* nrn_object_pop(void) {
+    // NOTE: the returned object should be unref'd when no longer needed
+    Object** obptr = hoc_objpop();
+    Object* new_ob_ptr = *obptr;
+    new_ob_ptr->refcount++;
```

Comment:
I'm therefore also suspicious about this refcount++. A function returning an existing object (vector does often) won't keep increasing the reference count?